File: ios/Classes/TiBLECharacteristicProxy.swift
/**
 * Appcelerator Titanium Mobile - Bluetooth Low Energy (BLE) Module
 * Copyright (c) 2020 by Axway, Inc. All Rights Reserved.
 * Proprietary and Confidential - This source code is not for redistribution
 */
import Foundation
import TitaniumKit
import CoreBluetooth

@objc
class TiBLECharacteristicProxy: TiProxy {
    private var _characteristic: CBCharacteristic!

    override init() {
        super.init()
    }

    convenience init(pageContext: TiEvaluator, characteristic: CBCharacteristic) {
        self.init()
        _init(withPageContext: pageContext)
        setCharacteristic(characteristic: characteristic)
    }

    func setCharacteristic(characteristic: CBCharacteristic) {
        _characteristic = characteristic
    }

    @objc
    func service() -> TiBLEServiceProxy? {
        guard let service = _characteristic.service else {
            return nil
        }

        return TiBLEServiceProxy(pageContext: self.pageContext, service: service)
    }
    @objc
    func properties() -> NSNumber {
        return NSNumber(value: _characteristic.properties.rawValue)
    }
    @objc
    func descriptors() -> [TiBLEDescriptorProxy] {
        guard let descriptors = _characteristic.descriptors else {
            return []
        }
        var objects = [TiBLEDescriptorProxy]()
        for descriptor in descriptors {
            objects.append(TiBLEDescriptorProxy(pageContext: pageContext, descriptor: descriptor))
        }
        return objects
    }
    @objc
    func value() -> TiBuffer? {
        guard let requestValue = _characteristic.value else {
            return nil
        }
        return TiBLEUtils.toTiBuffer(from: requestValue)._init(withPageContext: pageContext)
    }
    @objc
    func uuid() -> String {
        return _characteristic.uuid.uuidString
    }
    func characteristic() -> CBCharacteristic {
        return _characteristic
    }

    @objc
    func isNotifying() -> NSNumber {
        return NSNumber(value: _characteristic.isNotifying)
    }

    @objc
    func isMutable() -> NSNumber {
        if self is TiBLEMutableCharacteristicProxy {
            return NSNumber(true)
        }
        return NSNumber(false)
    }

    @objc(equal:)
    func equal(arg: Any?) -> NSNumber {
        guard let options = (arg as? [[String: Any]])?.first,
              let characteristic = options["characteristic"] as? TiBLECharacteristicProxy else {
            return NSNumber(false)
        }
        if self.uuid() == characteristic.uuid() {
            return NSNumber(true)
        }
        return NSNumber(false)
    }

}

File: android/src/appcelerator/ble/TiBLECharacteristicProxy.java
/**
 * Appcelerator Titanium Mobile - Bluetooth Low Energy (BLE) Module
 * Copyright (c) 2020 by Axway, Inc. All Rights Reserved.
 * Proprietary and Confidential - This source code is not for redistribution
 */
package appcelerator.ble;

import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import java.util.List;
import java.util.UUID;
import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollProxy;
import org.appcelerator.kroll.annotations.Kroll;
import ti.modules.titanium.BufferProxy;

@Kroll.proxy
public class TiBLECharacteristicProxy extends KrollProxy
{
	private BluetoothGattCharacteristic characteristic;

	public TiBLECharacteristicProxy(BluetoothGattCharacteristic characteristic)
	{
		this.characteristic = characteristic;
	}

	public BluetoothGattCharacteristic getCharacteristic()
	{
		return characteristic;
	}

	@Kroll.getProperty
	public TiBLEServiceProxy service()
	{
		return new TiBLEServiceProxy(characteristic.getService());
	}

	@Kroll.getProperty
	public String uuid()
	{
		return characteristic.getUuid().toString();
	}

	@Kroll.getProperty
	public BufferProxy value()
	{
		if (characteristic.getValue() == null) {
			return null;
		}
		return new BufferProxy(characteristic.getValue());
	}

	@Kroll.getProperty
	public int permissions()
	{
		return characteristic.getPermissions();
	}

	@Kroll.getProperty
	public int properties()
	{
		return characteristic.getProperties();
	}

	@Kroll.getProperty
	public TiBLEDescriptorProxy[] descriptors()
	{
		List<BluetoothGattDescriptor> descriptors = characteristic.getDescriptors();
		TiBLEDescriptorProxy[] descriptorProxies = new TiBLEDescriptorProxy[descriptors.size()];
		for (int i = 0; i < descriptors.size(); i++) {
			descriptorProxies[i] = new TiBLEDescriptorProxy(descriptors.get(i));
		}
		return descriptorProxies;
	}

	@Kroll.getProperty
	public boolean isMutable()
	{
		return false;
	}
}
